File: packages/scrape-kdl/test/parser.test.mjs
import assert from "node:assert/strict";
import { readFile } from "node:fs/promises";
import test from "node:test";
import { fileURLToPath } from "node:url";
import { parse } from "../dist/parser.js";

const root = fileURLToPath(new URL("../../..", import.meta.url));

test("raw multiline strings and nested comments match the documented KDL subset", () => {
  const source = `/* outer /* inner */ done */\nnode #"""\n  first\n  second\n  """# key=#"raw"#\n`;
  const { document, diagnostics } = parse("raw.kdl", source);
  assert.deepEqual(diagnostics, []);
  assert.equal(document.nodes.length, 1);
  assert.equal(document.nodes[0].arguments[0].value, "  first\n  second\n  ");
  assert.equal(document.nodes[0].properties[0].value.value, "raw");
});

test("slashdash suppresses nodes, arguments, properties, children, and typed components", () => {
  const source = `/- (ignored) dropped "node"\nkept /- (ignored) "drop" "first" /- old=1 new=2 /- { ignored } { real "ok" }\n`;
  const { document, diagnostics } = parse("slashdash.kdl", source);
  assert.deepEqual(diagnostics, []);
  assert.deepEqual(document.nodes.map((node) => node.name), ["kept"]);
  assert.deepEqual(document.nodes[0].arguments.map((value) => value.value), ["first"]);
  assert.deepEqual(document.nodes[0].properties.map((property) => [property.name, property.value.value]), [["new", 2]]);
  assert.deepEqual(document.nodes[0].children.map((node) => node.name), ["real"]);
});

test("integer bases, separators, floats, booleans, and null retain kinds and raw spellings", () => {
  const { document, diagnostics } = parse("numbers.kdl", "numbers 0x10 0o10 0b10 -0x2 1_000 1.25e+2 #true #false #null\n");
  assert.deepEqual(diagnostics, []);
  assert.deepEqual(document.nodes[0].arguments.map(({ kind, raw, value }) => ({ kind, raw, value })), [
    { kind: "int", raw: "0x10", value: 16 },
    { kind: "int", raw: "0o10", value: 8 },
    { kind: "int", raw: "0b10", value: 2 },
    { kind: "int", raw: "-0x2", value: -2 },
    { kind: "int", raw: "1_000", value: 1000 },
    { kind: "float", raw: "1.25e+2", value: 125 },
    { kind: "bool", raw: "#true", value: true },
    { kind: "bool", raw: "#false", value: false },
    { kind: "null", raw: "#null", value: null },
  ]);
});

test("signed 64-bit integers retain an exact internal value beyond Number safe range", () => {
  const { document, diagnostics } = parse("int64.kdl", "numbers 9223372036854775807 -9223372036854775808\n");
  assert.deepEqual(diagnostics, []);
  assert.deepEqual(document.nodes[0].arguments.map((value) => value.integerValue), [9223372036854775807n, -9223372036854775808n]);
});

test("quoted Unicode escapes and UTF-8 byte spans are deterministic", () => {
  const { document, diagnostics } = parse("unicode.kdl", "node \"é\\u{1f40e}\"\n");
  assert.deepEqual(diagnostics, []);
  const value = document.nodes[0].arguments[0];
  assert.equal(value.value, "é🐎");
  assert.deepEqual(value.span, {
    file: "unicode.kdl",
    start: { offset: 5, line: 1, column: 6 },
    end: { offset: 18, line: 1, column: 18 },
  });
});

test("unsupported representations and malformed tokens use stable parser diagnostics", () => {
  const annotated = parse("typed.kdl", "(type) node\n");
  assert.deepEqual(annotated.diagnostics, [{
    code: "E_TYPE_ANNOTATION_UNSUPPORTED",
    severity: "error",
    message: "KDL type annotations are not supported",
    span: { file: "typed.kdl", start: { offset: 0, line: 1, column: 1 }, end: { offset: 1, line: 1, column: 2 } },
  }]);

  const malformed = parse("bad.kdl", "node 0x @\n");
  assert.deepEqual(malformed.diagnostics.map((item) => [item.code, item.message, item.span.start.offset, item.span.end.offset]), [
    ["E_KDL_SYNTAX", "unexpected token \"0x\" in node", 5, 7],
    ["E_KDL_SYNTAX", "integer base prefix must be followed by a digit", 5, 7],
    ["E_KDL_SYNTAX", "unexpected token \"@\" in node", 8, 9],
    ["E_KDL_SYNTAX", "unexpected character '@'", 8, 9],
  ]);
});

test("the shared Go and TypeScript fuzz seeds and bounded mutations never throw or hang", async () => {
  const seeds = JSON.parse(await readFile(`${root}/fixtures/parser/fuzz-seeds.json`, "utf8"));
  for (const seed of seeds) assert.doesNotThrow(() => parse("fuzz.kdl", seed));
  let state = 0x5eed1234;
  const alphabet = "{}();=/#-_ abcdef0123456789\\\"\n\ré🐎";
  for (let sample = 0; sample < 2048; sample++) {
    state = xorshift(state);
    const length = state % 512;
    let source = "";
    for (let index = 0; index < length; index++) {
      state = xorshift(state);
      source += alphabet[state % alphabet.length];
    }
    assert.doesNotThrow(() => parse("fuzz.kdl", source));
  }
});

test("shared parser cases match Go acceptance and exact syntax diagnostics", async () => {
  const cases = JSON.parse(await readFile(`${root}/fixtures/parser/cases.json`, "utf8"));
  for (const testCase of cases) {
    const result = parse(`${testCase.name}.kdl`, testCase.source);
    assert.deepEqual(result.diagnostics, testCase.diagnostics, testCase.name);
  }
});

function xorshift(value) {
  value ^= value << 13;
  value ^= value >>> 17;
  value ^= value << 5;
  return value >>> 0;
}

File: internal/kdl/parser_test.go
package kdl

import "testing"

func TestParseRawMultilineString(t *testing.T) {
	source := []byte("extractor \"x\" version=\"2026-07-15\" language-version=\"2026-07-15\" {\n  source \"html\" {\n    fetch mode=\"browser\" url=\"https://example.invalid/\"\n    workflow { evaluate-js #\"\"\"\n      () => document.title\n      \"\"\"# }\n  }\n}\n")
	doc, diags := Parse("test.kdl", source)
	if diags.HasErrors() {
		t.Fatalf("unexpected diagnostics: %#v", diags)
	}
	if len(doc.Nodes) != 1 || len(doc.Nodes[0].Children) != 1 {
		t.Fatalf("unexpected document: %#v", doc)
	}
	workflow := doc.Nodes[0].Children[0].Children[1]
	js := workflow.Children[0]
	if got, ok := stringArgForTest(js, 0); !ok || got == "" {
		t.Fatalf("expected JavaScript string, got %#v", js.Arguments)
	}
}

func stringArgForTest(n *Node, idx int) (string, bool) {
	if idx >= len(n.Arguments) || n.Arguments[idx].Kind != ValueString {
		return "", false
	}
	s, ok := n.Arguments[idx].Value.(string)
	return s, ok
}

func TestDuplicatePropertiesArePreserved(t *testing.T) {
	doc, diags := Parse("test.kdl", []byte("extractor \"x\" version=\"2026-07-15\" version=\"2026-07-16\" language-version=\"2026-07-15\"\n"))
	if diags.HasErrors() {
		t.Fatalf("syntax parser must preserve duplicate properties for semantic validation: %#v", diags)
	}
	if got := len(doc.Nodes[0].PropertiesNamed("version")); got != 2 {
		t.Fatalf("expected two properties, got %d", got)
	}
}

func TestSlashdashSuppressesNodesEntriesAndChildren(t *testing.T) {
	source := []byte(`/-
ignored "node"
kept "first" /- "dropped-argument" "second" /- old=1 new=2 /- { ignored-child } { real-child "ok" }
`)
	doc, diags := Parse("test.kdl", source)
	if diags.HasErrors() {
		t.Fatalf("unexpected diagnostics: %#v", diags)
	}
	if len(doc.Nodes) != 1 {
		t.Fatalf("expected one retained node, got %d", len(doc.Nodes))
	}
	node := doc.Nodes[0]
	if node.Name != "kept" {
		t.Fatalf("node name = %q", node.Name)
	}
	if len(node.Arguments) != 2 {
		t.Fatalf("arguments = %#v", node.Arguments)
	}
	if got, _ := stringArgForTest(node, 0); got != "first" {
		t.Fatalf("first argument = %q", got)
	}
	if got, _ := stringArgForTest(node, 1); got != "second" {
		t.Fatalf("second argument = %q", got)
	}
	if len(node.Properties) != 1 || node.Properties[0].Name != "new" {
		t.Fatalf("properties = %#v", node.Properties)
	}
	if len(node.Children) != 1 || node.Children[0].Name != "real-child" {
		t.Fatalf("children = %#v", node.Children)
	}
}

func TestSlashdashCanSuppressTypedComponents(t *testing.T) {
	source := []byte(`/- (ignored) node "x"
kept /- (ignored) "drop" "keep"
`)
	doc, diags := Parse("test.kdl", source)
	if diags.HasErrors() {
		t.Fatalf("unexpected diagnostics: %#v", diags)
	}
	if len(doc.Nodes) != 1 || len(doc.Nodes[0].Arguments) != 1 {
		t.Fatalf("document = %#v", doc)
	}
	if got, _ := stringArgForTest(doc.Nodes[0], 0); got != "keep" {
		t.Fatalf("argument = %q", got)
	}
}

func TestParseKDLIntegerBases(t *testing.T) {
	doc, diags := Parse("test.kdl", []byte("numbers 0x10 0o10 0b10 -0x2\n"))
	if diags.HasErrors() {
		t.Fatalf("unexpected diagnostics: %#v", diags)
	}
	want := []int64{16, 8, 2, -2}
	if len(doc.Nodes) != 1 || len(doc.Nodes[0].Arguments) != len(want) {
		t.Fatalf("document = %#v", doc)
	}
	for i, expected := range want {
		if got, ok := doc.Nodes[0].Arguments[i].Value.(int64); !ok || got != expected {
			t.Fatalf("argument[%d] = %#v, want %d", i, doc.Nodes[0].Arguments[i].Value, expected)
		}
	}
}
